src/client/dfuse/pil4dfs/int_dfs.c

```diff
@@ -0,0 +1,4338 @@
+	}
+
+	if (oflags & O_CREAT) {
+		va_list arg;
+
+		va_start(arg, oflags);
+		mode = va_arg(arg, unsigned int);
+		va_end(arg);
+		two_args = 0;
+	}
+
+	if (!hook_enabled)
+		goto org_func;
+
+	/* Absolute path, dirfd is ignored */
+	if (path[0] == '/') {
+		if (two_args)
+			return open_common(libc_open, "new_openat", path, oflags);
+		else
+			return open_common(libc_open, "new_openat", path, oflags, mode);
+	}
+
+	/* Relative path */
+	idx_dfs = check_path_with_dirfd(dirfd, full_path, path);
+	if (idx_dfs >= 0) {
+		if (two_args)
+			return open_common(libc_open, "new_openat", full_path, oflags);
+		else
+			return open_common(libc_open, "new_openat", full_path, oflags, mode);
+	}
+
+org_func:
+	if (two_args)
+		return next_openat(dirfd, path, oflags);
+	else
+		return next_openat(dirfd, path, oflags, mode);
+}
+int
+openat64(int dirfd, const char *pathname, int oflags, ...) __attribute__((alias("openat")));
+
+int
+__openat_2(int dirfd, const char *path, int oflags)
+{
+	int  idx_dfs;
+	char full_path[DFS_MAX_PATH];
+
+	if (next_openat_2 == NULL) {
+		next_openat_2 = dlsym(RTLD_NEXT, "__openat_2");
+		assert(next_openat_2 != NULL);
+	}
+	if (!hook_enabled)
+		return next_openat_2(dirfd, path, oflags);
+
+	if (path[0] == '/')
+		return open_common(libc_open, "__openat_2", full_path, oflags);
+
+	idx_dfs = check_path_with_dirfd(dirfd, full_path, path);
+	if (idx_dfs >= 0)
+		return open_common(libc_open, "__openat_2", full_path, oflags);
+
+	return next_openat_2(dirfd, path, oflags);
+}
+
+int
+closedir(DIR *dirp)
+{
+	int fd;
+
+	if (next_closedir == NULL) {
+		next_closedir = dlsym(RTLD_NEXT, "closedir");
+		assert(next_closedir != NULL);
+	}
+	if (!hook_enabled)
+		return next_closedir(dirp);
+
+	if (!dirp) {
+		printf("dirp == NULL in closedir().\nQuit\n");
+		errno = EINVAL;
+		return (-1);
+	}
+
+	fd = dirfd(dirp);
+	if (fd >= FD_DIR_BASE) {
+		free_dirfd(dirfd(dirp) - FD_DIR_BASE);
+		return 0;
+	} else {
+		return next_closedir(dirp);
+	}
+}
+
+static struct dirent *
+new_readdir(DIR *dirp)
+{
+	int               rc    = 0;
+	struct DIRSTATUS *mydir = (struct DIRSTATUS *)dirp;
+
+	if (!hook_enabled || mydir->fd < FD_FILE_BASE)
+		return next_readdir(dirp);
+
+	if (mydir->fd < FD_DIR_BASE) {
+		printf("Error: invalid fd in readdir.\n");
+		errno = EINVAL;
+		return NULL;
+	}
+
+	if (mydir->num_ents)
+		goto out_readdir;
+
+	mydir->num_ents = READ_DIR_BATCH_SIZE;
+	while (!daos_anchor_is_eof(&mydir->anchor)) {
+		rc = dfs_readdir(dir_list[mydir->fd - FD_DIR_BASE].dfs_mt->dfs, mydir->dir_obj,
+				 &mydir->anchor, &mydir->num_ents, mydir->ents);
+		if (rc != 0)
+			goto out_null_readdir;
+
+		/* We have an entry, so return it */
+		if (mydir->num_ents != 0)
+			goto out_readdir;
+	}
+
+out_null_readdir:
+	mydir->num_ents = 0;
+	errno           = rc;
+	return NULL;
+out_readdir:
+	mydir->num_ents--;
+	mydir->offset++;
+	/*	mydir->ents[mydir->num_ents].d_ino = mydir->offset; */
+	return &mydir->ents[mydir->num_ents];
+}
+
+/**
```

Comment:
I can remove them now. Will put them back later after 2.4.

Reply:
oh well no need to remove them just because of release. im just asking if they are not needed to remove. if they are but just need more work, you can keep them.